beginner_source/text_sentiment_ngrams_tutorial.py

```diff
@@ -1,66 +1,131 @@
+#     >>> [475, 21, 30, 5286]
+#
+# Prepare the text processing pipeline with the tokenizer and vocabulary. The text and label pipelines will be used to process the raw data strings from the dataset iterators.
+
+
+text_pipeline = lambda x: vocab(tokenizer(x))
+label_pipeline = lambda x: int(x) - 1
+
+
+######################################################################
+# The text pipeline converts a text string into a list of integers based on the lookup table defined in the vocabulary. The label pipeline converts the label into integers. For example,
+#
+# ::
+#
+#     text_pipeline('here is the an example')
+#     >>> [475, 21, 2, 30, 5286]
+#     label_pipeline('10')
+#     >>> 9
+#
+
+
+
+######################################################################
+# Generate data batch and iterator 
+# --------------------------------
+#
+# `torch.utils.data.DataLoader <https://pytorch.org/docs/stable/data.html?highlight=dataloader#torch.utils.data.DataLoader>`__
+# is recommended for PyTorch users (a tutorial is `here <https://pytorch.org/tutorials/beginner/data_loading_tutorial.html>`__).
+# It works with a map-style dataset that implements the ``getitem()`` and ``len()`` protocols, and represents a map from indices/keys to data samples. It also works with an iterable datasets with the shuffle argumnet of ``False``.
+#
+# Before sending to the model, ``collate_fn`` function works on a batch of samples generated from ``DataLoader``. The input to ``collat_fn`` is a batch of data with the batch size in ``DataLoader``, and ``collate_fn`` processes them according to the data processing pipelines declared previouly. Pay attention here and make sure that ``collate_fn`` is declared as a top level def. This ensures that the function is available in each worker.
+#
+# In this example, the text entries in the original data batch input are packed into a list and concatenated as a single tensor for the input of ``nn.EmbeddingBag``. The offset is a tensor of delimiters to represent the beginning index of the individual sequence in the text tensor. Label is a tensor saving the labels of indidividual text entries.
+
+
+from torch.utils.data import DataLoader
 device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
+def collate_batch(batch):
+    label_list, text_list, offsets = [], [], [0]
+    for (_label, _text) in batch:
+         label_list.append(label_pipeline(_label))
+         processed_text = torch.tensor(text_pipeline(_text), dtype=torch.int64)
+         text_list.append(processed_text)
+         offsets.append(processed_text.size(0))
+    label_list = torch.tensor(label_list, dtype=torch.int64)
+    offsets = torch.tensor(offsets[:-1]).cumsum(dim=0)
+    text_list = torch.cat(text_list)
+    return label_list.to(device), text_list.to(device), offsets.to(device)    
+
+train_iter = AG_NEWS(split='train')
+dataloader = DataLoader(list(train_iter), batch_size=8, shuffle=True, collate_fn=collate_batch)
```

Comment:
I think we should split out ```list(train_iter)``` into its separate line and address it. Here the fact that the dataset is small makes it feasible to just materialize it into a list, but in general someone might want to use DataLoader for multiple streams (e.g. to apply tokenization and Vocab concurrently). Here the ```offset``` argument can help.

Reply:
Fixed the tutorial by passing an IterableDataset as an example.